### PR TITLE
Implements Enhancement #47

### PR DESCRIPTION
**Pull request by JD-Howard**
**Objective**
Should resolve the enhancement #47 request. Creates a context menu ability that adds Workspace files to an opened PRJ.

**Abstraction/Implementation**
I was just going to look at this because I was curious. It turned out to be an exceptionally easy update and recycled all existing functionality/localizations. Copied existing `addFile2Project()` command, added the arguments related to workspace context menu event and added an optional argument to the existing function that added files to project. Which created this new `addWorkspaceFile2Project()` command with next to no effort.

**Tests Performed**
Successfully added files to an open PRJ from the workspace, verified the contextual aspects weren't available when not applicable and verified that we couldn't accidentally add non-LSP files. Also verified our original Treeview version still functioned.

### CODE REVIEW

**Comment by hualin-adsk** [comment]
maybe it looks like this issue https://github.com/Autodesk-AutoCAD/AutoLispExt/issues/38 can also be covered?

**Comment by JD-Howard** [comment]
Yeah, that probably wouldn't be much more effort to do from the workspace. I wouldn't want to tack that onto this PR if we were talking about a new interface item within the Project Treeview though.

**Comment by hualin-adsk** [comment]
This also reminds me that some users want a tree view in project panel, and also want to add some non lisp files to project files (I will check old VLIDE if can support the non-lisp files, and check how the vlx compiler to treat those non-lisp files).

**Comment by JD-Howard** [comment]
The Treeview within the project panel is going to be a fairly large implementation change. The right click menu is currently filtered to only show up when an LSP file is selected and a PRJ is present. Removing half the filter isn't a big deal, but I'd have to do some research on the non-LSP file types. I know we have file type filtering at various locations, but I couldn't tell you what effects removing that could have.

**Comment by hualin-adsk** [comment]
I tried to compile the codes, but didn't find the context menu. And How about to click the context menu if there is no project file opened/created?

**Comment by JD-Howard** [comment]
If you have a project opened, then this should be working. You won't see the context menu items unless a PRJ is loaded. 

The last commit added the ability to process folders or even a mix of folders and files. If the folder contains non-LSP's, they were filtered out. I had to do this to work around the existing type checking I talked about earlier. IE, seems like our original implementation was specifically designed to abort the whole operation if it finds a non-LSP file.

**Comment by hualin-adsk** [comment]
For the non-lisp files in project issue, I tested in legacy VLIDE, and it looks like the `FAS compiler` can not support such case:
```
(VLISP-PROJECT-LIST
  :name
  aaa
  :own-list
  ("ab" "a" "..\\sam.dwg")
  :fas-directory
  nil
  :tmp-directory
  nil
  :project-keys
  (:BUILD (:standard))
  :context-id
  :AUTOLISP
)
;;; EOF
```

**Comment by JD-Howard** [comment]
I confirmed the same thing about the "other" file types. So, I would just close Issue #40 instead of creating a compatibility issue that probably won't add value.

**Comment by hualin-adsk** [comment]
I see from workspace now it is easy to add to project:
addfilefromworkspace

And I think we need to add a prompt messages after finished to add the file to project, for examples `"10 lisp files have been added to test.prj project"`

**Comment by JD-Howard** [comment]
Well, if we are going to deal with localizations on this, then maybe our workspace context menu should read "Add File(s) to Project"

I just recycled the old localization.

File: extension/src/project/projectCommands.ts [hunk 1]
```diff
@@ -99,6 +100,45 @@ export function registerProjectCommands(context: vscode.ExtensionContext) {
             }
         }));
 
+        context.subscriptions.push(vscode.commands.registerCommand('autolisp.addWorkspaceFile2Project', async (clickedFile: vscode.Uri, selectedFiles: vscode.Uri[]) => {
```
Comment (hualin-adsk) [inline extension/src/project/projectCommands.ts:103]: And I see in the project root node to right click there is also an menu "add file to project", for this menu it is better that it can also support to select a folder.

File: extension/src/project/projectCommands.ts [hunk 2]
```diff
@@ -99,6 +100,45 @@ export function registerProjectCommands(context: vscode.ExtensionContext) {
...
+            }
+            let selectedDirs = selectedFiles.filter(f => fs.statSync(f.fsPath, { bigint: false}).isDirectory());
+            selectedFiles = selectedFiles.filter(f => fs.statSync(f.fsPath, { bigint: false}).isFile());
+            
+            selectedDirs.forEach(dir => {
+                fs.readdirSync(dir.fsPath).forEach(name => {
+                    const fspath = dir.fsPath.replace(/[\/\\]$/, '') + '\\' + name;
+                    if (fs.existsSync(fspath) && fs.statSync(fspath, { bigint: false}).isFile()) {
+                        if (AutoLispExt.Documents.getSelectorType(fspath) === AutoLispExt.Selectors.lsp){
+                            selectedFiles.push(vscode.Uri.file(fspath));
```
Comment (hualin-adsk) [inline extension/src/project/projectCommands.ts:115]: Maybe it needs to test if the selectedFiles is null (the workspace has no lisp files)

File: extension/src/project/addFile2Project.ts [hunk 3]
```diff
@@ -3,14 +3,14 @@ import { ProjectTreeProvider, isFileAlreadyInProject, hasFileWithSameName } from
 import { AutoLispExt } from '../extension';
 import * as path from 'path';
 
-export async function AddFile2Project() {
+export async function AddFile2Project(fileList?: vscode.Uri[]) {
```
Comment (hualin-adsk) [inline extension/src/project/addFile2Project.ts:6]: the typescript new syntax is nice for null case

Comment (JD-Howard) [reply extension/src/project/addFile2Project.ts:6]: Yeah, the changes to this file do really show the value of the Typescript update. I didn't have to add any additional lines to recycle the method!

File: extension/src/project/projectCommands.ts (hunk 1 above)
Comment (JD-Howard) [reply extension/src/project/projectCommands.ts:103]: You don't think the ability to use standard multi-selection is sufficient? Does VSCode even have a decent folder picker? I feel like that's a pretty large deficiency existing in all newer Microsoft related frameworks.

File: extension/src/project/projectCommands.ts (hunk 2 above)
Comment (JD-Howard) [reply extension/src/project/projectCommands.ts:115]: I don't think this is even possible. From what I can tell, the context menu won't show up unless its on/over an LSP file or a Directory. If it is over a directory, then we have at least 1 URI even if it doesn't contain any files. IE, it just won't have anything to do work with.

File: package.json [hunk 4]
```diff
@@ -590,6 +599,18 @@
...
 				}
+			],
+			"explorer/context": [
+				{
+					"when": "resourceLangId == autolisp && autolisp.hasProject",					
+					"command": "autolisp.addWorkspaceFile2Project",
+					"group": "navigation@1"
+				},
+				{
+					"when": "explorerResourceIsFolder && autolisp.hasProject",
```
[the remark this answers is not in the document]
Comment (JD-Howard) [reply package.json:610]: This is, but I'll draw your attention Line 99 simply because it is where I recycled the previous localization that I am questioning should change to possibly pluralize the localized statement.

File: extension/src/project/projectCommands.ts [hunk 5]
```diff
@@ -99,6 +100,45 @@ export function registerProjectCommands(context: vscode.ExtensionContext) {
             }
         }));
 
+        context.subscriptions.push(vscode.commands.registerCommand('autolisp.addWorkspaceFile2Project', async (clickedFile: vscode.Uri, selectedFiles: vscode.Uri[]) => {
+            if (!selectedFiles || selectedFiles.length === 0 || getWarnIsSearching()){
+                return;
+            }
+            let selectedDirs = selectedFiles.filter(f => fs.statSync(f.fsPath, { bigint: false}).isDirectory());
+            selectedFiles = selectedFiles.filter(f => fs.statSync(f.fsPath, { bigint: false}).isFile());
```
Comment (hualin-adsk) [inline extension/src/project/projectCommands.ts:108]: Your codes cthere support to select a dir. But in the project node it doesn't support to select dir. I think to keep consisstene in hte project node context menu it is best to have this.

Comment (JD-Howard) [reply extension/src/project/projectCommands.ts:108]: Okay, but I am not very good at designing icons. So, which SVG's would you like associated with this new interface element?

Comment (hualin-adsk) [reply extension/src/project/projectCommands.ts:108]: I only mean to refine the functionality of the menu item to support selecting dir, see pic:
addfilefromworkspace-menu

File: extension/src/project/addFile2Project.ts [hunk 6]
```diff
@@ -90,3 +90,17 @@ async function SelectLspFiles() {
     if (fileUris && fileUris.length > 0)
         return Promise.resolve(fileUris);
 }
+
+
+export async function SelectFolder(): Promise<vscode.Uri[]> {
+    let label = AutoLispExt.localize("autolispext.project.addfolder.openlabel", "Add Folder");
```
Comment (hualin-adsk) [inline extension/src/project/addFile2Project.ts:96]: Thanks for your patience!

Sorry I didn't make me understood on this item. 
Please see the GIF: 
https://github.com/Autodesk-AutoCAD/AutoLispExt/pull/51#discussion_r528318730

And I mean that we only improve the "add file to project" menu item and make it support to select a folder.

**Comment by Rina-Y** [comment]
Does this work for individual files or multiple files selected inside the workspace rather than just folders? Was just curious due to the watching your screencap that it only had folders being added.

**Comment by JD-Howard** [comment]
> 
> 
> Does this work for individual files or multiple files selected inside the workspace rather than just folders? Was just curious due to the watching your screencap that it only had folders being added.

If your in a workspace, and you Shift+Select multiple files, all the files/folders in that selection would get added to the project. Note that this will not recursively dig through sub-folders; just the top level files.

**Review by hualin-adsk — APPROVED** [review]

**Comment by hualin-adsk** [comment]
It is best that adding a prompt message for how many files have been added to project file.

**Comment by JD-Howard** [comment]
I just put the quantity message back in from commit before the revert. I also added File Type filtering for *.LSP's since we've now determined PRJ's only want to work with LSP's and a workspace presents a lot of opportunity for non-LSP's to get in the way.
